crates/packet_inspector/src/context.rs

```diff
@@ -0,0 +1,365 @@
+                    Ok(None) => return "S2cPlayPacket".to_string(),
+                    Err(err) => return format!("{:?}", err),
+                };
+                format!("{pkt:#?}")
+            }
+        }
+    }
+}
+
+pub struct Logger {
+    pub include_filter: Option<Regex>,
+    pub exclude_filter: Option<Regex>,
+}
+
+pub enum ContextMode {
+    Gui(egui::Context),
+    Cli(Logger),
+}
+
+pub struct Context {
+    pub mode: ContextMode,
+    pub last_packet: AtomicUsize,
+    pub selected_packet: RwLock<Option<usize>>,
+    pub(crate) packets: RwLock<Vec<Packet>>,
+    pub(crate) packet_count: RwLock<usize>,
+    pub filter: RwLock<String>,
+    c2s_style: Style,
+    s2c_style: Style,
+}
+
+impl Context {
+    pub fn new(mode: ContextMode) -> Self {
+        Self {
+            mode,
+            last_packet: AtomicUsize::new(0),
+            selected_packet: RwLock::new(None),
+            packets: RwLock::new(Vec::new()),
+            filter: RwLock::new("".into()),
+            packet_count: RwLock::new(0),
+
+            c2s_style: Style::new().green(),
+            s2c_style: Style::new().purple(),
+        }
+    }
+
+    pub fn clear(&self) {
+        self.last_packet.store(0, Ordering::Relaxed);
+        *self.selected_packet.write().expect("Poisoned RwLock") = None;
+        self.packets.write().expect("Poisoned RwLock").clear();
+        if let ContextMode::Gui(ctx) = &self.mode {
+            ctx.request_repaint();
+        }
+    }
+
+    pub fn add(&self, mut packet: Packet) {
+        match &self.mode {
+            ContextMode::Gui(ctx) => {
+                packet.id = self.last_packet.fetch_add(1, Ordering::Relaxed);
+                self.packets.write().expect("Poisoned RwLock").push(packet);
+                ctx.request_repaint();
+            }
+            ContextMode::Cli(logger) => {
+                if let Some(include_filter) = &logger.include_filter {
+                    if !include_filter.is_match(&packet.packet_name) {
+                        return;
+                    }
+                }
+                if let Some(exclude_filter) = &logger.exclude_filter {
+                    if exclude_filter.is_match(&packet.packet_name) {
+                        return;
+                    }
+                }
+
+                let arrow = match &packet.direction {
+                    PacketDirection::ClientToServer => "↑",
+                    PacketDirection::ServerToClient => "↓",
+                };
+
+                let style = match &packet.direction {
+                    PacketDirection::ClientToServer => self.c2s_style,
+                    PacketDirection::ServerToClient => self.s2c_style,
+                };
+
+                println!(
+                    "[{}] ({}) {}",
+                    systemtime_strftime(packet.created_at),
+                    arrow.style(style),
+                    packet.get_packet_string_no_format().style(style)
+                )
```

Comment:
When redirecting the output to a file, the color escape codes are still present. You can fix this with [atty](https://crates.io/crates/atty). Would be nice to have a command line argument to override the default behavior.